src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -0,0 +1,348 @@
+        {
+            // TODO: this one will be implemented later
+            // https://github.com/dotnet/roslyn/issues/30165
+            var hierarchy = hierarchyContent.Hierarchy;
+            if (hierarchy == null)
+            {
+                // solution
+                return Task.FromResult(false);
+            }
+
+            var itemId = hierarchyContent.ItemId;
+
+            if (hierarchy.GetCanonicalName(itemId, out var path) == 0)
+            {
+                var attr = File.GetAttributes(path);
+                if (attr.HasFlag(FileAttributes.Directory))
+                {
+                    // directory
+                }
+                else
+                {
+                    // document
+                }
+            }
+
+            return Task.FromResult(false);
+        }
+
+        private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBufferScope, ICodeCleanUpExecutionContext context, CancellationToken cancellationToken)
+        {
+            using (var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(context.OperationContext.UserCancellationToken, cancellationToken))
+            {
+                cancellationToken = cancellationTokenSource.Token;
+
+                var buffer = textBufferScope.SubjectBuffer;
+                if (buffer != null)
+                {
+                    using (var scope = context.OperationContext.AddScope(allowCancellation: true, description: EditorFeaturesResources.Applying_changes))
+                    {
+                        var progressTracker = new ProgressTracker((description, completed, total) =>
+                        {
+                            if (scope != null)
+                            {
+                                scope.Description = description;
+                                scope.Progress.Report(new ProgressInfo(completed, total));
+                            }
+                        });
+
+                        var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                        var codeCleanupService = document.GetLanguageService<ICodeCleanupService>();
+
+                        // TODO: enable all diagnostics for now, need to be replace by inclusion/ exclusion list from .editorconfig
+                        // https://github.com/dotnet/roslyn/issues/30163
+                        var allDiagnostics = codeCleanupService.GetAllDiagnostics();
+
+                        var enabedDiagnosticSets = ArrayBuilder<DiagnosticSet>.GetInstance();
+
+                        foreach (var diagnostic in allDiagnostics.Diagnostics)
+                        {
+                            foreach (var diagnosticId in diagnostic.DiagnosticIds)
+                            {
+                                if (context.EnabledFixIds.IsFixIdEnabled(diagnosticId))
```

Comment:
this is strange. if we have 1:1 map from diagnosticId to fixId, then why we have DiagnosticSet to group them? might be to map description to show in progress bar?

Reply:
yeah, this mapping is strange. Let's improve it post Preview1.